File: html.go
// html.go is part of the Go models package.
//
// @author R. S. Doiel, <rsdoiel@caltech.edu>
//
// Copyright (c) 2024, Caltech
// All rights not granted herein are expressly reserved by Caltech.
//
// Redistribution and use in source and binary forms, with or without modification, are permitted provided
// that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and 
//    the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions
//    and the following disclaimer in the documentation and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or
//    promote products derived from this software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, 
// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
package models

import (
	"fmt"
	"io"
	"strings"
)

// ModelToHTML takes a model and renders an input form. The form is not
// populated with values through that could be done easily via JavaScript and DOM calls.
func ModelToHTML(out io.Writer, model *Model) error {
	// Include the description as an HTML comment.
	// Write opening form element
	if model.Id != "" {
		fmt.Fprintf(out, "<!-- %s: %s -->\n", model.Id, model.Description)
		fmt.Fprintf(out, "<form id=%q", model.Id)
	} else {
		fmt.Fprintf(out, "<!-- %s -->\n", model.Description)
		fmt.Fprintf(out, "<form")
	}
	for k, v := range model.Attributes {
		switch k {
		case "checked":
			fmt.Fprintf(out, " checked")
		case "required":
			fmt.Fprintf(out, " required")
		default:
			fmt.Fprintf(out, " %s=%q", k, v)
		}
	}
	cssBaseClass := strings.ReplaceAll(strings.ToLower(model.Id), " ", "_")
	fmt.Fprintf(out, ">\n")
	for _, elem := range model.Elements {
		ElementToHTML(out, cssBaseClass, elem)
	}
	if !model.HasElementType("submit") {
		cssName := fmt.Sprintf("%s-submit", cssBaseClass)
		fmt.Fprintf(out, `  <div class=%q><input class=%q type="submit" value="submit"> <input class=%q type="reset" value="cancel"></div>`,
			cssName, cssName, cssName)
	}

	// Write closing form element
	fmt.Fprintf(out, "\n</form>\n")
	return nil
}

// ElementToHTML renders an individual element as HTML, includes label as well as input element.
func ElementToHTML(out io.Writer, cssBaseClass string, elem *Element) error {
	cssClass := fmt.Sprintf("%s-%s", cssBaseClass, strings.ToLower(elem.Id))
	fmt.Fprintf(out, "  <div class=%q>", cssClass)
	switch strings.ToLower(elem.Type) {
	case "textarea":
		if elem.Label != "" {
			if name, ok := elem.Attributes["name"]; ok {
				fmt.Fprintf(out, "<label class=%q set=%q>%s</label> <textarea class=%q", cssClass, name, elem.Label, cssClass)
			} else {
				fmt.Fprintf(out, "<label class=%q set=%q>%s</label> <textarea class=%q name=%q", cssClass, elem.Id, elem.Label, cssClass, elem.Id)
			}
		} else {
			fmt.Fprintf(out, "<textarea class=%q", cssClass)
		}
	case "button":
		fmt.Fprintf(out, "<button class=%q", cssClass)
	default:
		if elem.Label != "" {
			if name, ok := elem.Attributes["name"]; ok {
				fmt.Fprintf(out, "<label class=%q set=%q>%s</label> <input class=%q type=%q", cssClass, name, elem.Label, cssClass, elem.Type)
			} else {
				fmt.Fprintf(out, "<label class=%q set=%q>%s</label> <input class=%q name=%q type=%q", cssClass, elem.Id, elem.Label, cssClass, elem.Id, elem.Type)
			}
		} else {
			fmt.Fprintf(out, "<input class=%q type=%q", cssClass, elem.Type)
		}
	}
	if elem.Id != "" {
		fmt.Fprintf(out, " id=%q", elem.Id)
	}
	for k, v := range elem.Attributes {
		switch k {
		case "checked":
			fmt.Fprintf(out, " checked")
		case "required":
			fmt.Fprintf(out, " required")
		default:
			fmt.Fprintf(out, " %s=%q", k, v)
		}
	}
	switch strings.ToLower(elem.Type) {
	case "button":
		fmt.Fprintf(out, " >%s</button>", elem.Label)
	case "textarea":
		fmt.Fprintf(out, " ></textarea>")
	default:
		fmt.Fprintf(out, " >")
	}
	fmt.Fprintf(out, "</div>\n")
	return nil
}

File: html.ts
import { Model } from "./model.ts";
import { Element } from "./element.ts";

// ModelToHTML takes a model and renders an input form. The form is not
// populated with values, but that could be done easily via JavaScript and DOM calls.
export async function modelToHTML(out: WritableStream, model: Model): Promise<void> {
  const writer = out.getWriter();
  const encoder = new TextEncoder();

  // Include the description as an HTML comment.
  // Write opening form element
  if (model.id !== "") {
    await writer.write(encoder.encode(`<!-- ${model.id}: ${model.description} -->\n`));
    await writer.write(encoder.encode(`<form id="${model.id}"`));
  } else {
    await writer.write(encoder.encode(`<!-- ${model.description} -->\n`));
    await writer.write(encoder.encode(`<form`));
  }

  for (const [k, v] of Object.entries(model.attributes)) {
    switch (k) {
      case "checked":
        await writer.write(encoder.encode(" checked"));
        break;
      case "required":
        await writer.write(encoder.encode(" required"));
        break;
      default:
        await writer.write(encoder.encode(` ${k}="${v}"`));
    }
  }

  const cssBaseClass = model.id.toLowerCase().replaceAll(" ", "_");
  await writer.write(encoder.encode(">\n"));

  for (const elem of model.elements) {
    await elementToHTML(writer, encoder, cssBaseClass, elem);
  }

  if (!model.hasElementType("submit")) {
    const cssName = `${cssBaseClass}-submit`;
    await writer.write(encoder.encode(`  <div class="${cssName}"><input class="${cssName}" type="submit" value="submit"> <input class="${cssName}" type="reset" value="cancel"></div>\n`));
  }

  // Write closing form element
  await writer.write(encoder.encode("</form>\n"));
  await writer.close();
}

// ElementToHTML renders an individual element as HTML, including the label as well as the input element.
async function elementToHTML(writer: WritableStreamDefaultWriter, encoder: TextEncoder, cssBaseClass: string, elem: Element): Promise<void> {
  const cssClass = `${cssBaseClass}-${elem.id.toLowerCase()}`;
  await writer.write(encoder.encode(`  <div class="${cssClass}">`));

  // Generate a default label if not provided
  const labelText = elem.label || elem.id.charAt(0).toUpperCase() + elem.id.slice(1);
  const name = elem.attributes["name"] || elem.id;
  await writer.write(encoder.encode(`<label class="${cssClass}" for="${name}">${labelText}</label> `));

  switch (elem.type.toLowerCase()) {
    case "textarea":
      await writer.write(encoder.encode(`<textarea class="${cssClass}"`));
      break;
    case "button":
      await writer.write(encoder.encode(`<button class="${cssClass}"`));
      break;
    default:
      await writer.write(encoder.encode(`<input class="${cssClass}" type="${elem.type}"`));
  }

  if (elem.id !== "") {
    await writer.write(encoder.encode(` id="${elem.id}"`));
  }

  for (const [k, v] of Object.entries(elem.attributes)) {
    switch (k) {
      case "checked":
        await writer.write(encoder.encode(" checked"));
        break;
      case "required":
        await writer.write(encoder.encode(" required"));
        break;
      default:
        await writer.write(encoder.encode(` ${k}="${v}"`));
    }
  }

  switch (elem.type.toLowerCase()) {
    case "button":
      await writer.write(encoder.encode(`>${elem.label}</button>`));
      break;
    case "textarea":
      await writer.write(encoder.encode("></textarea>"));
      break;
    default:
      await writer.write(encoder.encode(">"));
  }

  await writer.write(encoder.encode("</div>\n"));
}
